### ISSUE TITLE
Adapting the repo to my specific problem

### ISSUE THREAD

**Issue by VVIERV00** [MarcoMeter/recurrent-ppo-truncated-bptt#3]
Hi! Your repo is a very good guideline to LSTM w/ PPO. I have been trying to adapt it to my personal gym env. 
My env:

- Obs: Dict {time-series-data: 60 last entries (like a window moving each timestep) of 15 features : 60x15 + 3 1d vectors}
- Actions: Box(2,)
- Aprox steps by ep: 400

What I think to use as a solution:

- LTSM (just for the time-series data) + Linear (fed with the previous result flattened + rest vectors of the obs) and then feed the actor critics policies of PPO. 

Problems and doubts:

1. You have proposed to flatten the 60x15 and then use attention on that input. Can you provide a simple code example? I haven't worked with attention and there is not much info out there.
2. Could I use LTSM w Attention? or maybe dual encoder-decoder LTSM with or without Attention?. I want a simple but powerful approach, I don't think that the agent has to work on 'predicting' the next values on the time series (which as in stocks are difficult to predict), maybe it can figure out trends changes so that it must adapt to those with actions turning on or off triggers on the env. I wonder what arch can suit this better. 
3. Right now I am trying to adapt your work to my env. I have an error and is that _train_mini_batch gives me this error Expected hidden[0] size (3, 1260, 128), got [3, 21, 128] when feeding the model. The issue is that recurrent_mini_batch_generator only collects states at the beginning of a seq. I don't know how to handle it as I have already made some changes in shapes that may not be correct and I feel very new to this. What do I use as batches? fixed secuences of obs? I'm messing up on how to feed the obs to the hole learning architecture. Any insight will be appreciated :)

**Comment by MarcoMeter** [comment]
Hi @VVIERV00 

1.

I haven't worked with attention yet, but it is on my TODO. So I cannot give further advice on that. I'd look at the implementation that [Unity ml-agents achieved for attention](https://github.com/Unity-Technologies/ml-agents/blob/main/ml-agents/mlagents/trainers/torch/attention.py). However, ml-agents is hard to get into as it is complex framework.

I'm not sure if I understood your time-series input right. 60x15 (I assume sequence length is 60) is the data that the agent receives as observation, correct?
If so, you pretty much have to build an encoder for this to process that input in your model. The data needs also to be stored inside the buffer. The hidden state (if your encoder uses an LSTM) does not need to be tracked inside the buffer.

3.

If my assumption on your observation space is correct, you don't need the recurrent part of the PPO policy. Like I said, you'd add an encoder to the model, potentially using an LSTM, that encodes your time series data. The result of that is concatenated to your other observation data. Then the rest of the network propagates the data as usual to estimate the value and the policy.

**Comment by VVIERV00** [comment]
Hi @MarcoMeter 

1. Yeah the framework seems rather complex but useful, thanks.
2. By sequence I mean (maybe I'm wrong) that those 60 sequences are in order, every step adding the last reading and removing the trailing one on the seq (every step a new 1x15 data is loaded but I also add the last 59 entries to help the agent decide better. Hereby I attach an example of 4x6 (as a shorthand of 60x15) to make it clear. The rest of the observations are flat vecs bc I don't find previous information useful so just add the last reading on those other obs.
![IMG_20210907_120737__01](https://user-images.githubusercontent.com/17448761/132329492-8aa00b63-2cf1-4cb6-8eb3-8842941fae99.jpg)

3. I don't understand, the only recurrent part of the PPO is the LSTM itself isn't it? What is the shape of the encoded result? Let me show you what my model looks like, I am not sure about the hiperparams of the lstm. Can you explain about the buffer: do I store the raw data or encoded one? Why can I forget about the hidden state? So, in general I have doubts on how to use the buffer in training time.
[model.txt](https://github.com/MarcoMeter/recurrent-ppo-truncated-bptt/files/7120788/model.txt)

**Comment by MarcoMeter** [comment]
I'm not really sure if I understood your observations correctly.

Can't you flatten the entire observation data and use it as a single vector observation? Then you can just proceed without any changes on the recurrent policy.

**Comment by VVIERV00** [comment]
Yes I could flatten the 60x15, but I wonder it would become less clear to the agent that all it has are 15 features in 60 different time steps. You recommend flattening the time-series and feed (just this obs, or maybe the rest vecs also?) to the LSTM?

**Comment by MarcoMeter** [comment]
`60x15 + 15 =  915` are a lot of features to derive a policy from. Theoretically, the approach of flattening everything and then feeding the model is still correct. Practically, there are many possibilities to improve the encoding of the these observation features.

Your problem is basically an encoding problem. The recurrent layer of this repository's model is not meant to encode the observations (Of course this can be argued as the observations are fed directly to the LSTM. A linear layer between the vector observation and LSTM is recommendable). So you have to find a way to derive a suitable representation from your observation data.

This is how I imagine your model's architecture:

![model](https://user-images.githubusercontent.com/6996955/132335177-e981009d-c7cd-45f1-8271-2d36b7d78bd9.png)

**Comment by VVIERV00** [comment]
The model architecture fits pretty much what I am trying to achieve.

1. In your diagram, do I need to keep track of the internal states of the encoder for training time? (in case it's an LSTM)
2. In the case I use an LSTM encoder, do I really need the later LSTM layer?
3. About the number of features being a lot, using encoder + autoencoder (e.g. LTSM bidirectional with Attention) won't help true?
4.  I am seduced to use LTSM bidirectional with Attention, maybe I could use it after the linear layer. 
5. If the use of the encoders is reducing the size of the features, what's a general rule for the output size? from 1x900 to let say 1x200?

**Comment by MarcoMeter** [comment]
1a. All what you need to change is the interface between every component and of course the modules and the forward pass of the model. By interface I mean that the environments supplies the time series data, that needs to be fed to the model. This data needs to be stored inside the buffer as well.

1b. Maybe a better and essentially easier strategy for you is to flatten the time-series data and concatenate it with your regular vector obs. Then you wouldn't need to make changes to the interface. Inside your model code you would revert that transformation.

2. It depends on your use-case. At least that layer can be easily turned on and off.

3. and 4. I don't have experiences myself on this matter. That's what you have to figure out for yourself.

5. There is no golden rule for that. It depends on your data/task and so on. My gut feeling would end up in a range from 256 to 512.
